### PR TITLE
Feature/provision correct buckets

### LINKED ISSUE #656: Provision correct buckets based on sg config
In progress on https://github.com/couchbaselabs/mobile-testkit/tree/feature/provision_correct_buckets, but running into this error now;

https://gist.github.com/tleyden/ef71dc273e1b53dfcb30163be320713c

### PR DESCRIPTION
**Pull request by tleyden**
Fixes #656.

This change is [Reviewable](https://reviewable.io/reviews/couchbaselabs/mobile-testkit/657)

### CODE REVIEW

**Comment by tleyden** [comment]
jenkins job in progress: http://uberjenkins.sc.couchbase.com/view/QE%20Dev/job/syncgateway-functional-tests-dev/139/console

**Comment by tleyden** [comment]
Fixes #656

**Comment by tleyden** [comment]
kicked off a bigger jenkins job run, 11/11 tests passed: http://uberjenkins.sc.couchbase.com/view/QE%20Dev/job/syncgateway-functional-tests-dev/141

**Comment by tleyden** [comment]
I ran the full suite and there were 41 failures, which seems like a lot

http://uberjenkins.sc.couchbase.com/view/QE%20Dev/job/syncgateway-functional-tests-dev/142/robot/

**Comment by tleyden** [comment]
Re-running against the mobile-testkit master branch for comparison: 

http://uberjenkins.sc.couchbase.com/view/QE%20Dev/job/syncgateway-functional-tests-dev/143/

**Comment by tleyden** [comment]
I ran the full functional tests against the mobile-testkit master branch:

http://uberjenkins.sc.couchbase.com/view/QE%20Dev/job/syncgateway-functional-tests-dev/143/parameters/

and there were also 41 failures, so this should be safe to merge

**Comment by sethrosetter** [comment]
@tleyden I ran this branch with the QE job (should be green) and noticed quite a few failures.  http://uberjenkins.sc.couchbase.com/job/centos7-mobile-sync-gateway-functional-tests/394. We need to resolved them before it is merge-able.

**Comment by tleyden** [comment]
In [these logs](http://uberjenkins.sc.couchbase.com/job/centos7-mobile-sync-gateway-functional-tests/394/robot/report/log.html#s1-s3-s1-t2)

```
>>> Waiting for server to be in 'healthy' state

  sync_gateway configuration
  ------------------------------------------
  version:          1.3.1
  build number:     16
  commit:           None
  config path:      resources/sync_gateway_configs/sync_gateway_default_functional_tests_di.json
  build flags:      
  skip bucketcreation: False	
00:56:40.660	FAIL	ValueError: Expecting property name: line 1 column 1047 (char 1046)
```

**Comment by tleyden** [comment]
Ran into an issue trying to call `install_sync_gateway.py` with an sg config that was trying to provision a walrus bucket:

```
DELETE http://ec2-52-207-217-222.compute-1.amazonaws.com:8091/pools/default/buckets/grocery-sync 200
DELETE http://ec2-52-207-217-222.compute-1.amazonaws.com:8091/pools/default/buckets/sgload 200
DELETE http://ec2-52-207-217-222.compute-1.amazonaws.com:8091/pools/default/buckets/todolite 200
Traceback (most recent call last):
  File "libraries/provision/install_sync_gateway.py", line 213, in <module>
    install_sync_gateway(sync_gateway_install_config)
  File "libraries/provision/install_sync_gateway.py", line 98, in install_sync_gateway
    create_server_buckets(os.environ["CLUSTER_CONFIG"], sync_gateway_config)
  File "libraries/provision/install_sync_gateway.py", line 140, in create_server_buckets
    bucket_names = get_buckets_from_sync_gateway_config(sync_gateway_config.config_path)
  File "libraries/provision/install_sync_gateway.py", line 148, in get_buckets_from_sync_gateway_config
    config = Config(config_path_full)
  File "/Users/tleyden/Development/mobile-testkit/libraries/testkit/config.py", line 31, in __init__
    self.discover_bucket_name_set(conf_obj)
  File "/Users/tleyden/Development/mobile-testkit/libraries/testkit/config.py", line 59, in discover_bucket_name_set
    bucket_names_from_config.append(val["bucket"])
KeyError: 'bucket'
```

**Comment by tleyden** [comment]
Pushed more commits and kicked off functional test:

http://uberjenkins.sc.couchbase.com/job/centos7-mobile-sync-gateway-functional-tests/403/

**Comment by tleyden** [comment]
Full func test run in progress: http://uberjenkins.sc.couchbase.com/job/centos7-mobile-sync-gateway-functional-tests/424/

One test was passing / failing sporadically: http://uberjenkins.sc.couchbase.com/job/centos7-mobile-sync-gateway-functional-tests/423/

**Comment by sethrosetter** [comment]
Reviewed 6 of 12 files at r1.
Review status: 3 of 15 files reviewed at latest revision, 1 unresolved discussion.

---

*[keywords/CouchbaseServer.py, line 183 at r5](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS3ZsXEln5W5bxA41MK:-KS3ZsXEln5W5bxA41ML:b-tbo1ci) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/f6aac5d8dd2066ab606be8ae06e4a686223388e9/keywords/CouchbaseServer.py#L183)):*
> ```Python
>         """
>         resp = requests.get("{}/pools/default".format(url), auth=self._auth)
>         resp_json = resp.json()
> ```

We should raise exception on status

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657)*

**Comment by tleyden** [comment]
Review status: 3 of 15 files reviewed at latest revision, 1 unresolved discussion.

---

*[keywords/CouchbaseServer.py, line 183 at r5](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS3ZsXEln5W5bxA41MK:-KS3c_hWeIK75hwVB7xu:bu15oag) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/f6aac5d8dd2066ab606be8ae06e4a686223388e9/keywords/CouchbaseServer.py#L183)):*
<details><summary><i>Previously, sethrosetter (Seth Rosetter) wrote…</i></summary>
> We should raise exception on status
</details>
Done.  Pushed a new commit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657)*

**Comment by tleyden** [comment]
@sethrosetter the full test run passed: http://uberjenkins.sc.couchbase.com/job/centos7-mobile-sync-gateway-functional-tests/424/

**Comment by sethrosetter** [comment]
Reviewed 4 of 12 files at r1, 3 of 4 files at r2, 1 of 2 files at r3, 1 of 3 files at r4, 2 of 3 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[keywords/CouchbaseServer.py, line 194 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS7R14okJ8mVvuwUWxA:-KS7R14okJ8mVvuwUWxB:b1x4qm6) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/keywords/CouchbaseServer.py#L194)):*
> ```Python
>         """
>         if len(bucket_names) == 0:
>             return
> ```

When would we expect this scenario. Should we be strict and throw an exception?

---

*[libraries/provision/test_install_sync_gateway.py, line 7 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS7RV8iqcKdav-KNfGZ:-KS7RV8iqcKdav-KNfG_:b3ksb) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/libraries/provision/test_install_sync_gateway.py#L7)):*
> ```Python
> 
> ```

`

---

*[libraries/provision/test_install_sync_gateway.py, line 8 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS7RW-cMuAtLgHGOvXW:-KS7RW-cMuAtLgHGOvXX:b-2coevu) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/libraries/provision/test_install_sync_gateway.py#L8)):*
> ```Python
> 
> 
> class TestInstallSyncGateway(unittest.TestCase):
> ```

Should probably be pytest format but pytest can run `unittest` tests as well

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657)*

File: libraries/provision/provision_cluster.py [hunk 1]
```diff
@@ -13,11 +13,14 @@
...
+from keywords.utils import log_info
+
 def provision_cluster(couchbase_server_config, sync_gateway_config):
 
     test_output("\n>>> Cluster info:\n")
 
-    with open(os.environ["CLUSTER_CONFIG"], "r") as ansible_hosts:
+    cluster_config = os.environ["CLUSTER_CONFIG"]
+    with open(cluster_config, "r") as ansible_hosts:
         test_output(ansible_hosts.read())
```
Comment (sethrosetter) [inline libraries/provision/provision_cluster.py]: These should be `log_info`

File: libraries/provision/provision_cluster.py [hunk 2]
```diff
@@ -46,15 +49,19 @@ def provision_cluster(couchbase_server_config, sync_gateway_config):
...
     # Install nginx
-    install_nginx(os.environ["CLUSTER_CONFIG"])
+    install_nginx(cluster_config)
+
+    log_info(">>> Done provisioning cluster...")
+
 
-    test_output(">>> Done provisioning cluster...")
 
 def test_output(output):
```
Comment (sethrosetter) [inline libraries/provision/provision_cluster.py]: This can be removed

**Comment by sethrosetter** [comment]
Filed - https://github.com/couchbaselabs/mobile-testkit/issues/683 to get the unit tests for the framework in CI

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-:-KS7W-jLUEJD87BU33lD:bqgtbe2)*

**Comment by tleyden** [comment]
Review status: 13 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[keywords/CouchbaseServer.py, line 194 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS7R14okJ8mVvuwUWxA:-KS7oVAFX3O1xxhLtub3:bx5uplm) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/keywords/CouchbaseServer.py#L194)):*
<details><summary><i>Previously, sethrosetter (Seth Rosetter) wrote…</i></summary>
> When would we expect this scenario. Should we be strict and throw an exception?
</details>
If the SG config is "walrus only", this will happen.  It's totally normal, no need to raise an exception.

---

*[libraries/provision/provision_cluster.py, line 24 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KQXVXDRfFH9s1fLgIaP-r6-24:-KS7qJY_RyCIvQBz6xAs:b-1l530p) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/libraries/provision/provision_cluster.py#L24)):*
<details><summary><i>Previously, sethrosetter (Seth Rosetter) wrote…</i></summary>
> These should be `log_info`
</details>
Done.  Pushed new commit.

---

*[libraries/provision/provision_cluster.py, line 66 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KQXVXDRfFH9s1fLgIaP-r6-66:-KS7qLal9wAH6syCuNAV:b-yf5lyh) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/libraries/provision/provision_cluster.py#L66)):*
<details><summary><i>Previously, sethrosetter (Seth Rosetter) wrote…</i></summary>
> This can be removed
</details>
Done.  Pushed new commit

---

*[libraries/provision/test_install_sync_gateway.py, line 8 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS7RW-cMuAtLgHGOvXW:-KS7qkqo4RVDoOmc18KL:b-fy49l2) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/libraries/provision/test_install_sync_gateway.py#L8)):*
> Should probably be pytest format but pytest can run unittest tests as well

I added a TODO comment for now.  Will do later whenever updating that test and re-verifying

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657)*

**Comment by tleyden** [comment]
Review status: 13 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[libraries/provision/test_install_sync_gateway.py, line 7 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS7RV8iqcKdav-KNfGZ:-KS7qtyENxHd5Z-rqUlr:b-896fix) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/libraries/provision/test_install_sync_gateway.py#L7)):*
<details><summary><i>Previously, sethrosetter (Seth Rosetter) wrote…</i></summary>
> `
</details>
Done.

---

*[libraries/provision/test_install_sync_gateway.py, line 8 at r6](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657#-KS7RW-cMuAtLgHGOvXW:-KS7quLxp2FFyz1p6Gj5:b-896fix) ([raw file](https://github.com/couchbaselabs/mobile-testkit/blob/daf83ed9559d4e11f98692bfd470ba4284a4bbe5/libraries/provision/test_install_sync_gateway.py#L8)):*
<details><summary><i>Previously, tleyden (Traun Leyden) wrote…</i></summary>
> > Should probably be pytest format but pytest can run unittest tests as well
> 
> I added a TODO comment for now.  Will do later whenever updating that test and re-verifying
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/couchbaselabs/mobile-testkit/657)*

**Pull request merged**